### PR TITLE
Vectorized (rebased on master)

### PR DESCRIPTION
**Pull request by bqpd**
@mayork, @mjburton11, try this branch; it's `feassweep` (https://github.com/hoburg/gpkit/pull/925) without the commits not related to vectorization.

### CODE REVIEW

**Comment by mayork** [comment]
@bqpd will you be around after the group meeting? The engine code still won't run on this branch, but I have commits on master where it does run

**Comment by mayork** [comment]
@bqpd the negine still won't run on this (it actually won't run on any of hte master comitts after https://github.com/hoburg/gpkit/commit/66d9606549ce5d0f3e949de21c892f68b8383ebd)..can we discuss post group meeting?

**Comment by bqpd** [comment]
Sure!

**Comment by mjburton11** [comment]
I'm working on `1682master` with `gasmaleperf.py`.  For some reason any value of `W_{pay}` between 9.7-10.6 returns the dreaded `UNKNOWN` with `mosek`.  But it works up to `W_{pay} = 20` and down to `W_{pay} = 0.5`.

**Comment by bqpd** [comment]
Michael, I got an email about this comment but can't find it on the
PR...did it get resolved?

**Comment by mjburton11** [comment]
It was resolved.  It turned out to be a modeling issue so I moved it here: https://github.com/hoburg/gpkit-models/issues/95

In the future I'll be sure to reference something if I move it.

File: gpkit/tests/t_model.py [hunk 1]
```diff
@@ -448,13 +448,42 @@ def __init__(self, n, **kwargs):
...
+    def __init__(self, box):
+        A_wall = Variable("A_{wall}", 100, "m^2", "Upper limit, wall area")
+        A_floor = Variable("A_{floor}", 50, "m^2", "Upper limit, floor area")
+
+        constraints = [2*box["h"]*box["w"] + 2*box["h"]*box["d"] <= A_wall,
+                       box["w"]*box["d"] <= A_floor]
+
+        Model.__init__(self, None, constraints)
+
+
```
Comment (whoburg) [inline gpkit/tests/t_model.py]: @arthur-brown, fyi, this commit was inspired by your simple box example that fails on master but works on this branch

Comment (whoburg) [inline gpkit/tests/t_model.py]: (f06a4cf88c55b71abbe3344ed67b2c3a4d1d0799)

**Comment by whoburg** [comment]
hey @bqpd I added a few commits that you may want to look over; I'm beginning a quick intermediate code review now

File: gpkit/__init__.py [hunk 2]
```diff
@@ -95,6 +96,50 @@ def __call__(self, arg):
...
+    num = MODELNUM_LOOKUP[model]
+    MODELNUMS.append(num)
+    MODELNUM_LOOKUP[model] += 1
+    return num, (tuple(MODELS), tuple(MODELNUMS))
+
+
+def end_variable_naming():
+    global MODELS, MODELNUMS
+    MODELS = MODELS[:-1]
+    MODELNUMS = MODELNUMS[:-1]
```
Comment (whoburg) [inline gpkit/__init__.py]: are `MODELS` and `MODELNUMS` lists, or not? It's confusing that they start out being assigned as `[]` and get appended to by `begin_variable_naming`, but then get assigned to the last element here. What happens if we then enter `begin_variable_naming` again (which will try to append) ?

File: gpkit/__init__.py [hunk 3]
```diff
@@ -95,6 +96,50 @@ def __call__(self, arg):
...
+
+class NamedVariables(object):
+    def __init__(self, model):
+        self.model = model
+
+    def __enter__(self):
+        begin_variable_naming(self.model)
+
+    def __exit__(self, type_, val, traceback):
+        end_variable_naming()
```
Comment (whoburg) [inline gpkit/__init__.py]: this class appears unused in the code base except for one occurrence in `t_model`. Do we still need it?

File: gpkit/__init__.py [hunk 4]
```diff
@@ -95,6 +96,50 @@ def __call__(self, arg):
...
+
+
+class vectorize(object):
```
Comment (whoburg) [inline gpkit/__init__.py]: should we capitalize?

File: gpkit/__init__.py [hunk 5]
```diff
@@ -95,6 +96,50 @@ def __call__(self, arg):
 
 enable_units()
 
+VECTORIZATION = ()
+MODELS = []
+MODELNUMS = []
+MODELNUM_LOOKUP = defaultdict(int)
+
+
+def begin_variable_naming(model):
```
Comment (whoburg) [inline gpkit/__init__.py]: docstrings needed throughout new functions and classes introduced in `__init__`

File: gpkit/constraints/model.py [hunk 6]
```diff
@@ -38,24 +35,29 @@ class Model(CostedConstraintSet):
     `program` is set during a solve
     `solution` is set at the end of a solve
     """
-    _nums = defaultdict(int)
+
     name = None
     num = None
+    naming = None
```
Comment (whoburg) [inline gpkit/constraints/model.py]: maybe an inline comment to disambiguate `name` and `naming`?

File: gpkit/constraints/model.py [hunk 7]
```diff
@@ -38,24 +35,29 @@ class Model(CostedConstraintSet):
...
     program = None
     solution = None
 
-    def __init__(self, cost=None, constraints=None,
-                 substitutions=None, name=None):
+    def __new__(cls, *args, **kwargs):
+        name = kwargs.pop("name", None)
+        obj = super(Model, cls).__new__(cls, *args, **kwargs)
+        if not name:
+            name = cls.name if cls.name else cls.__name__
```
Comment (whoburg) [inline gpkit/constraints/model.py]: is this better implemented as `getattr(cls, "name", cls.__name__)`?

File: gpkit/constraints/model.py [hunk 8]
```diff
@@ -38,24 +35,29 @@ class Model(CostedConstraintSet):
...
         CostedConstraintSet.__init__(self, cost, constraints, substitutions)
-        if self.__class__.__name__ != "Model" and not name:
-            name = self.__class__.__name__
-        if name:
-            self.name = name
-            self.num = Model._nums[name]
-            Model._nums[name] += 1
-            self._add_modelname_tovars(self.name, self.num)
+        if self.name:
+            end_variable_naming()
```
Comment (whoburg) [inline gpkit/constraints/model.py]: heh, yikes 😮

File: gpkit/constraints/relax.py [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+            descr.pop("value", None)
+            descr["units"] = "-"
+            relaxation = Variable(**descr)
+            self.relaxvars.append(relaxation)
+            relaxation_constraints.append([value/relaxation <= original,
+                                           original <= value*relaxation,
+                                           relaxation >= 1])
+        Model.__init__(self, NomialArray(self.relaxvars).prod(),
+                       [constraints, relaxation_constraints])
+        self.substitutions = substitutions
```
Comment (whoburg) [inline gpkit/constraints/relax.py]: these seem unrelated to this PR. Shall we split them off to a separate PR?

File: gpkit/nomials/variables.py [hunk 10]
```diff
@@ -160,3 +181,13 @@ def __new__(cls, shape, *args, **descr):
         obj.key = VarKey(**obj.descr)
 
         return obj
+
+
+class VectorizableVariable(Variable):
```
Comment (whoburg) [inline gpkit/nomials/variables.py]: inheritance is tricky here, should it double-inherit from `Variable` and `ArrayVariable`?

File: gpkit/tests/t_model.py [hunk 11]
```diff
@@ -386,7 +387,8 @@ def test_small_named_signomial(self):
         nonzero_adder = 0.1  # TODO: support reaching zero, issue #348
         with SignomialsEnabled():
             J = 0.01*(x - 1)**2 + nonzero_adder
-            m = Model(z, [z >= J], name="SmallSignomial")
+            with NamedVariables("SmallSignomial"):
+                m = Model(z, [z >= J])
```
Comment (whoburg) [inline gpkit/tests/t_model.py]: as far as I can tell, only use of `NamedVariables` in all of gpkit

File: gpkit/tests/t_vars.py [hunk 12]
```diff
@@ -4,6 +4,7 @@
 from gpkit import (Monomial, NomialArray, Variable, VarKey,
                    VectorVariable, ArrayVariable)
 import gpkit
+from gpkit.nomials import Variable as RealVariable
```
Comment (whoburg) [inline gpkit/tests/t_vars.py]: let tweak the name so it doesn't sound like 'real' as in a real number?

**Comment by whoburg** [comment]
@mayork reports that #946 is also an issue with this branch, which has been causing some problems / hindering progress

**Comment by bqpd** [comment]
@whoburg, @mayork, #946 should now be fixed on this branch

File: gpkit/tests/t_model.py (hunk 11 above)
Comment (bqpd) [inline gpkit/tests/t_model.py]: Yeah, the thought was that users might want it? But I think you're right, I should remove it.

File: gpkit/__init__.py (hunk 4 above)
Comment (bqpd) [inline gpkit/__init__.py]:
```python
with Vectorize(3):
    ...

# vs 

with vectorize(3):
    ...
```

File: gpkit/__init__.py (hunk 2 above)
Comment (bqpd) [inline gpkit/__init__.py]: They're not assigned to the last elements here; on the contrary, `[:-1]` just chops off the last element.

Comment (bqpd) [inline gpkit/__init__.py]: I'll use `pop()` instead to make this clearer.

File: gpkit/nomials/variables.py (hunk 10 above)
Comment (bqpd) [inline gpkit/nomials/variables.py]: Hmm, sure!

File: gpkit/tests/t_model.py (hunk 11 above)
Comment (bqpd) [inline gpkit/tests/t_model.py]: yeah, unless we switch to using `setup` methods...

File: gpkit/constraints/relax.py (hunk 9 above)
Comment (bqpd) [inline gpkit/constraints/relax.py]: done!

File: gpkit/__init__.py (hunk 3 above)
Comment (bqpd) [inline gpkit/__init__.py]: Potentially, if we return to `setup` methods.

File: gpkit/tests/t_vars.py (hunk 12 above)
Comment (bqpd) [inline gpkit/tests/t_vars.py]: changed to `PlainVariable`

File: gpkit/constraints/model.py (hunk 7 above)
Comment (bqpd) [inline gpkit/constraints/model.py]: shortened considerably

File: gpkit/constraints/model.py (hunk 6 above)
Comment (bqpd) [inline gpkit/constraints/model.py]: done!

**Comment by whoburg** [comment]
@1ozturkbe, is #946 fixed for you on this branch now?

File: gpkit/__init__.py (hunk 2 above)
Comment (whoburg) [inline gpkit/__init__.py]: doh, gotcha. yep, pop() will be clearer.

File: gpkit/__init__.py (hunk 3 above)
Comment (whoburg) [inline gpkit/__init__.py]: ugh, is that a possibility? ⛔️

File: gpkit/constraints/model.py (hunk 8 above)
Comment (whoburg) [inline gpkit/constraints/model.py]: hm, sounds like this might warrant a discussion in person

File: gpkit/__init__.py (hunk 5 above)
Comment (bqpd) [inline gpkit/__init__.py]: done!

File: gpkit/__init__.py [hunk 13]
```diff
@@ -115,6 +116,58 @@ def __call__(self, arg):
...
 
+VECTORIZATION = []
+MODELS = []
+MODELNUMS = []
+MODELNUM_LOOKUP = defaultdict(int)
+
+
+def begin_variable_naming(model):
+    "Appends a model name and num to the environment."
+    MODELS.append(model)
```
Comment (whoburg) [inline gpkit/__init__.py]: yay, you realized that `global` was not needed here?

**Comment by whoburg** [comment]
this is looking great to me; I think we still need

1. confirmation that #946 is resolved on this branch
2. documentation of these features on RTD

**Comment by mayork** [comment]
you now can't say `from gpkit import vectorize`...not sure if the init was changed last night but I think it's nice to be able to make the import as simple as possible

**Comment by bqpd** [comment]
Oops, it's now capitalized

**Comment by bqpd** [comment]
from gpkit import Vectorize

**Comment by bqpd** [comment]
Need to fix #950 on this branch before merging.

**Comment by mayork** [comment]
@1ozturkbe so is this running the d8 code now?

**Comment by 1ozturkbe** [comment]
Yes it is; Ned was hoping to come find you and deliver the good news!

**Comment by bqpd** [comment]
retest this please

**Comment by bqpd** [comment]
@galbramc, Reynolds seems quite sad right now...

**Comment by galbramc** [comment]
reynolds had a harddrive failure. I'm still working on getting everything back that is needed for testing.

**Comment by galbramc** [comment]
test this please

**Comment by bqpd** [comment]
Alack, poor reynolds! Here's to a speedy recovery :beers:

**Comment by bqpd** [comment]
superseded by #953
